lib/base/span.h

```diff
@@ -0,0 +1,144 @@
+// ready. All access to span is facilitated through free functions to ease
+// transition to std::span in the future.
+//
+// IsEmpty(), GetSize(), GetByteSize() and MakeSpan() provide support for common
+// interactions with Span.
+//
+// Example usage:
+//
+// std::vector<double> mutable_data(50);
+// const std::array<double, 10> immutable_data;
+//
+// const auto mutable_span = base::MakeSpan(mutable_data);
+// const auto immutable_span = base::MakeSpan(immutable_data);
+//
+// double* mutable_ptr = GetBase(mutable_span);
+// const double* immutable_ptr = GetBase(immutable_span);
+//
+// assert(50 == GetSize(mutable_span));
+// assert(!IsEmpty(mutable_span));
+// assert(50 * sizeof(double) == GetByteSize(mutable_span));
+//
+template <typename T>
+class Span {
+ public:
+  constexpr Span() = default;
+
+  // Construct a new span using 'base' and 'limit'.
+  // An explicitly constructed span must never have 'base' == nullptr. Users
+  // should prefer other approaches like std::optional<Span<T>> to represent a
+  // nullable span.
+  constexpr Span(T* const base, T* const limit) : base_(base), limit_(limit) {
+    INVARIANT_NE(nullptr, base);
+    INVARIANT_NE(nullptr, limit);
+    INVARIANT_LE(base, limit);
+  }
+
+  // Default copy constructor.
+  constexpr Span(const Span&) = default;
+
+  // Default copy assignment operator.
+  Span& operator=(const Span&) = default;
+
+  // Default move constructor.
+  constexpr Span(Span&&) = default;
+
+  // Default move assignment operator.
+  Span& operator=(Span&&) = default;
+
+  // Implicit conversion from Span<T> to Span<const T>.
+  constexpr operator Span<const T>() const {
+    return Span<const T>(base_, limit_);
+  }
+
+ private:
+  // Return a pointer to the first element in 'span'.
+  friend constexpr T* GetBase(const Span span) { return span.base_; }
+
+  // Return a pointer to an address one position past the last element in
+  // 'span'.
+  friend constexpr T* GetLimit(const Span span) { return span.limit_; }
+
+  // Pointer to the first element in this span.
+  T* base_ = nullptr;
+
+  // Pointer to an address one position past the last element in this span.
+  // An empty span will always have 'base_' == 'limit_'.
+  T* limit_ = nullptr;
+};
+
+// Return true iff 'span' contains no elements.
+template <typename T>
+inline constexpr bool IsEmpty(const Span<T> span) {
+  return GetLimit(span) == GetBase(span);
+}
+
+// Return the number of elements in 'span'.
+template <typename T>
+inline constexpr size_t GetSize(const Span<T> span) {
+  return GetLimit(span) - GetBase(span);
+}
+
+// Return the size of 'span' in bytes.
+template <typename T>
+inline constexpr size_t GetByteSize(const Span<T> span) {
+  return sizeof(T) * GetSize(span);
+}
+
+// Return a new span starting at 'base' and containing 'size' elements.
+template <typename T>
+inline constexpr Span<T> MakeSpan(T* const base, const size_t size) {
+  return Span<T>(base, base + size);
+}
```

Comment:
The stated intent is to fill in until the C++20 implementation arrives, but this implementation exposes a vastly different API, based on standalone instead of member functions. The concern is twofold:

1) What happens when C++20 comes along and other standard library APIs start depending on std::span. You can argue that only new code would be affected, but then you have a mix of APIs for doing the same thing in the code base.
2) As a community project, we should be adopting the APIs that will be familiar to others.

It should be noted that there are at least 3 other implementations of std::span available today with varying levels of cost and complexity. Would it perhaps be better to adopt one of these?:

a) https://github.com/microsoft/GSL/tree/master/include/gsl (probably not, since it does more than span and pulls in a bunch of deps, but I believe this implementation was the basis of the standardization proposal)
b) https://github.com/martinmoene/span-lite
c) https://github.com/tcbrindle/span/tree/master/include/tcb

some of which are more efficient in terms of implementation too (only storing size when needed, etc).

Reply:
I intend to use only a fraction of the built in functionality present in std::span so I do not wish to add a new dependency or implement our own fully functional std::span due to the complexity that brings to the project. In terms of efficiency, you have a good point that in cases with statically known size std::span does require less storage, however those are not likely to be relevant in our application (Packets and eBPF programs are unlikely to have statically known size).

I do intend to extend Span with more helper functions that are not present on the C++20 std::span. I would also prefer that they have syntax similar to the core functions like GetSize() which means that I will prefer functional style wrappers when possible.

Perhaps it is a mistake to bring std::span into this discussion. My intent was to reference the ideas present in std::span and that I would use the standard library implementation if it was available for "free". Lacking an available and standard implementation I would prefer that our implementation did not have dangerous features like operator[], .front() or .back() that fail badly on empty spans.

I consider Span to be a value and I would prefer to treat it as such. I don't write 10.subtract(5), so why should I write my_span.first(5). If it would help, I'm happy to add more helpers to clarify the intended usage. I was hoping to add these as needed when motivating examples are implicitly available.